Rerunning Staging pipeline One Pager (#13463)
dotnet/arcade: Documentation/TeamProcess/One-Pagers/retrying-stages-staging-pipeline.md

# Retrying Stages to Address Errors in Staging pipeline

## What is the current process?

Today we must run the entire Staging pipeline if we encounter any errors in the pipeline. 
For eg: If required validation fails then we will have to run Prep, Signing and then Validation stage. 

> There is no way to test the pipeline with fewer stages. Since most of the later stages are dependent on earlier stages that takes a lot of time to comeplete. 

Reviewer: ```suggestion
There is no way to test the pipeline with fewer stages. Since most of the later stages are dependent on earlier stages that takes a lot of time to complete. 
```

Author: updated

> There is no way to test the pipeline with fewer stages. Since most of the later stages are dependent on earlier stages that takes a lot of time to comeplete. 

Reviewer: "There is no way to test the pipeline...", "If we want to test..." In my opinion, the aim is not focused on testing, but rather to enhance the UX and speed up the resolution of pipeline issues during 'production usage'.

Author: This section is more about describing the current process than list what we want to achieve out of this epic. 

I agree with the aim, I have mentioned this under the advantages section.

| 4. Split the pipeline into 2 <li> Anything that alters the artifacts ( Signing) <li>Validation and publishing | <li> .Net 9 we are moving towards moving signing (anything that alters/creates the build artifacts is going to be moved) to main builds, so by splitting this we will align to work with .Net 9 ( which is awesome!) <li> Making the pipeline lot less complex<li> We can introduce testing infra (inject DI in the second pipeline, publish to test vs prod containers) <li> Makes the whole pipeline simpler <li> If we want to fix something in the second pipeline we don't have to rerun the first one. So fixes are and flaky test reruns are lot faster and easier. | <li> Have to maintain 2 pipeline <li> BCM work has to be reevaluated.| 

## Goal and Motivation 

After carefully analyzing the pros and cons of all the above approaches, we decided to go with splitting the pipeline. 

The proposal here is to split the Staging Pipeline into two different pipelines and add the ability to rerun stages. Thinking of this work as 2 part process.
#### Version 1 (V1):

> 1. First pipeline will contain anything the alters the artifacts (Eg: Signing, SBOM generation). 

Reviewer: ```suggestion
1. First pipeline will contain anything that alters the artifacts (Eg: Signing, SBOM generation). 
```

Author: Updated

  start --> vs_insertion["VS Insertion \n Ring \n <b>~50m</b>"]
  vs_insertion --> vs_insertion_override["VS Insertion \n Override \n <b></b>"]
  vs_insertion_override --> test_team_sign_off["Waiting for \n Test team \n to Sign off \n <b></b>"]
  test_team_sign_off --> staging_ring["Staging Ring \n <b></b>"]
  staging_ring --> staging_ring_override["Staging Ring \n Override \n <b></b>"]
  staging_ring_override --> sign_off_for_finalizing_release["Sign off for \nfinalizing \n the release \n <b></b>"]
  sign_off_for_finalizing_release --> finalize_staging_ring["Finalize \n Staging Ring \n <b></b>"]
  sign_off_for_finalizing_release --> publish_cit_validated_assets["Publish CTI \n validated assets \n <b>1h20m</b>"]
  finalize_staging_ring --> approve_publishing_dotnetcs_internal["Approve \n publishing to \n dotnetcsinternal \n <b></b>"]
  approve_publishing_dotnetcs_internal --> handoff_publishing_ring["Handoff \n Publishing Ring \n (dotnetcsinternal) \n <b></b>"]
```

#### Risks: 
1. We are dissecting the staging pipeline. We need to make sure all the assets are published to the correct storage containers. 

> 1. Verify if asset publishing is happening to the right container.

Reviewer 2: Some of these aren't stated as risks. Can you restate them in that form, with the issues that you have identified as mitigations for those risks?

Author: Yeah, I didn't know what to do with them. They are not really risk but I want to state them.

Reviewer 2: I think the risk is mainly (paraphased) "we're doing work to staging, we might screw some publishing up".

Author: I will just add them under Additional information or something like that..

Author: > I think the risk is mainly (paraphased) "we're doing work to staging, we might screw some publishing up".

I missed this response. Will update thanks.,

# Retrying Stages to Address Errors in Staging pipeline

## What is the current process?

Today we must run the entire Staging pipeline if we encounter any errors in the pipeline. 

> For eg: If required validation fails then we will have to run Prep, Signing and then Validation stage. 

Reviewer 2: This isn't a great example because it doesn't alter artifacts or push to endpoints, and can be overridden if necessary. A better example might be a case where a package push fails, we need to rerun everything to fix the bug.

## Problems with current process:
1.	Rerunning a single stage in the pipeline is not possible without executing the entire pipeline again.
1.	Running only the stage where changes were implemented is not feasible.
1.	Re-executing the entire process and testing consumes a considerable amount of time.
1.  Staging pipeline is complicated to understand, we need a simpler process.

## Different approach for staging pipeline: 

| Approach | Pros | Cons | 
|----------|------|-------|
| 1. We add some parameters in the pipeline, and make staging pipeline download the pipeline assets uploaded in the Signing stage of build with build id 12345 in the new pipeline run. Since the assets are already signed we just run required validation and publishing stage only We will need build id of the successful build. We will use azure apis to download the pipeline artifacts |   <li>  Easier to test     <li> We can pick and choose which stages to rerun                                                         <li>Relatively easy to implement <li> Release pipeline is doing something similar | <li> Seems primitive and hacky | 

> | 2. Pipeline should be intelligent enough, like if we rerun the pipeline with the same BarId, it should look up the data in timeline builds and if the stage was previously successful from previous runs it should just not do anything. Like here if prep ring was successful, so it would skip prep and go to signing. Signing was already successful so now it will skip this stage and just download assets from the build with build if 12345. Then eventually just run Required Validation and publishing stage | <li> Lot of code change in every stage <li> Timeline data is in Kusto and they sometime take time to show up in kusto so prone to errors(waiting can help) | <li> Need to capture details which build ran with which barid <li> Seems hacky | 

Reviewer 2: I also don't like this approach from the standpoint of predictability. It adds some non-determinism (based on time, availability of systems, etc.).

> | 3. Create a cache to cache the Signed assets, so next time we re run the pipeline for the same barID, We look for signed assets in the cache. The assets are hashed, so if there is no change in the bits, we do not have to run that stage. So in case required validation was previously successful for the signed assets hash  then skip this stage.| <li> Lot cleaner | <li> We may be adding  complexity to our current infra <li> Have to maintain cache | 

Reviewer 2: We also have to build in an override if we explicitly need to re-sign. This also doesn't save a ton of time in 7.0/8.0, where in-build signing reduces the time in the signing stage.

Author: added,

| 4. Split the pipeline into 2 <li> Anything that alters the artifacts ( Signing) <li>Validation and publishing | <li> .Net 9 we are moving towards moving signing (anything that alters/creates the build artifacts is going to be moved) to main builds, so by splitting this we will align to work with .Net 9 ( which is awesome!) <li> Making the pipeline lot less complex<li> We can introduce testing infra (inject DI in the second pipeline, publish to test vs prod containers) <li> Makes the whole pipeline simpler <li> If we want to fix something in the second pipeline we don't have to rerun the first one. So fixes are and flaky test reruns are lot faster and easier. | <li> Have to maintain 2 pipeline <li> BCM work has to be reevaluated.| 

## Goal and Motivation 

After carefully analyzing the pros and cons of all the above approaches, we decided to go with splitting the pipeline. 

The proposal here is to split the Staging Pipeline into two different pipelines and add the ability to rerun stages. Thinking of this work as 2 part process.
#### Version 1 (V1):

> 1. First pipeline will contain anything that alters the artifacts (Eg: Signing, SBOM generation). 

Reviewer 2: ```suggestion
1. First pipeline will contain anything that alters or generates artifacts (Eg: Signing, SBOM generation). 
```

1. Second pipeline will contain Validation and publishing to various storage accounts.
#### Version 2 (V2): 

> 1. Ability to add rerun stages in the Second pipeline.

Reviewer 2: I think the re-runnng of stages in the second pipeline won't require much, if any, work. You can select which stages you want to run when you launch, so if you're making fixes and rerunning sections of the second pipeline. There should be minimal critical dependencies between stages in the second pipeline, since no stage will generate artifacts for use later. The inter-dependencies should (mostly) just be approvals.

Author: Yeah I agree with this. We can use the basic functionality of skipping selected stages.

Author: Changed it. Closing this

  vs_insertion --> vs_insertion_override["VS Insertion \n Override \n <b></b>"]
  vs_insertion_override --> test_team_sign_off["Waiting for \n Test team \n to Sign off \n <b></b>"]
  test_team_sign_off --> staging_ring["Staging Ring \n <b></b>"]
  source_code_validation_override --> staging_ring["Staging Ring \n <b>~1h10m</b>"]
  staging_ring --> staging_ring_override["Staging Ring \n Override \n <b></b>"]
  staging_ring_override --> sign_off_for_finalizing_release["Sign off for \nfinalizing \n the release \n <b></b>"]
  sign_off_for_finalizing_release --> finalize_staging_ring["Finalize \n Staging Ring \n <b></b>"]
  sign_off_for_finalizing_release --> publish_cit_validated_assets["Publish CTI \n validated assets \n <b>1h20m</b>"]
  finalize_staging_ring --> approve_publishing_dotnetcs_internal["Approve \n publishing to \n dotnetcsinternal \n <b></b>"]
  approve_publishing_dotnetcs_internal --> handoff_publishing_ring["Handoff \n Publishing Ring \n (dotnetcsinternal) \n <b></b>"]
  sbom_generation --> sbom_generation_override[SBOM Generation Override]
```

## V1:
### Splitting Staging Pipeline:

Proposed implementation is splitting the pipeline into two pipelines.

> #### 1. First pipeline: Stage-Dotnet-Sign-Artifacts

Reviewer 2: ```suggestion
#### 1. First pipeline: Stage-Dotnet-Prepare-Artifacts
```

Reviewer 2: We sign, generate SBOMs, etc.

Author: done

The Stage-Dotnet-Sign-Artifacts pipeline will contain stages that alters the artifacts. 
```mermaid
flowchart LR;
  prep["Prep Ring \n <b>~30min</b>"] --> prep_override[Prep Ring Override]
  prep_override --> signing["Signing Ring \n <b>~50min</b>"]
  signing--> sbom_generation["SBOM Generation \n <b>~20m</b>"]
  signing--> kick_2nd_pipeline["Stage-Dotnet-Validate-Publish pipeline kick off \n <b>~20m</b>"]

>   sbom_generation --> sbom_generation_override[SBOM Generation Override]

Reviewer 2: kick_2nd_pipeline may need to depend on sbom_generation. I think we'll eventually end up with some assets generation (vsix packages for workloads) that need to include the sboms.

But for now this is fine.

Author: Changed it to after sbom_generation stage

```
#### 2. Second pipeline: Stage-Dotnet-Validate-Publish
The Stage-Dotnet-Validate-Publish pipeline will contain validation and publishing stages.
```mermaid
flowchart LR;

>   downloadAssets["Download Signed Assets \n <b>~30min</b>"] --> validation["Validation <b>~5h</b>"]

Reviewer 2: What's the purpose of downloadAssets stage? Wouldn't this just be a step within the validation, etc. jobs?

Author: we don't need it, removed it. Also I don't have a better way to put the three stages together, it kinda looks orphaned with no starting stage so I added an additional stage..

<img width="667" alt="image" src="https://github.com/dotnet/arcade/assets/47157394/201daca2-4c8e-4212-9bc0-13537eb6975b">

Author: in reality we don't have "start" stage

  sign_off_for_finalizing_release --> finalize_staging_ring["Finalize \n Staging Ring \n <b></b>"]
  sign_off_for_finalizing_release --> publish_cit_validated_assets["Publish CTI \n validated assets \n <b>1h20m</b>"]
  finalize_staging_ring --> approve_publishing_dotnetcs_internal["Approve \n publishing to \n dotnetcsinternal \n <b></b>"]
  approve_publishing_dotnetcs_internal --> handoff_publishing_ring["Handoff \n Publishing Ring \n (dotnetcsinternal) \n <b></b>"]
```

### Advantages of splitting pipeline: 
1. In .Net 9 we are going to move signing to main build. By splitting pipeline we are going to be in alignment with that plan, meaning we can retire the first pipeline when time comes and only the second pipeline will be staging pipeline then. 
1. Bug fix and testing in validation and publishing stages are lot faster. We do not have to wait for the build to be signed everytime we make a fix to validation/publishing stage or re-run flaky tests.
1. We can add ability to rerun to smaller subset of stages as compared to Stage-Dotnet pipeline.

### Interface between the Stage-Dotnet-Sign-Artifacts Pipeline and Stage-Dotnet-Validate-Publish pipeline:

> The first pipeline Stage-Dotnet-Sign-Artifacts (May be during the create SBOM stage or after) kicks off a build in the Stage-Dotnet-Validate-Publish pipeline. This is similar to what we have in maestro promotion pipeline. The Stage-Dotnet-Sign-Artifacts is not dependent on the Stage-Dotnet-Validate-Publish pipeline to be completed.

Reviewer 3: Have we considered having Signing Validation in the first pipeline? In my opinion this makes more sense, the first pipeline prepares and validates everything and hands it to the pipeline that stages everything. Also, I might be wrong here, but I think that if we have some signing issues, for example some runtime files are not signed, we'd have to work on the Signing.props in the runtime repo, flow that new build to installer and start the Staging pipeline again, meaning that we'd have to run the Gather-Drop and all those stages again. This makes me believe that Signing Validation should belong to the first pipeline

Author: I think we should run signing validation in parallel to all other validation. Reason being we already have a pipeline that runs signing validation daily. If there is a problem with signing we would anyway catch it early from the Validation pipeline. 

Also the whole idea of splitting pipeline into 2 is to reduce the time and make it less complex. I feel it will defeat that purpose if we do that, unless we start the second pipeline in parallel as soon as the signing happens then singing validation and other validation still happens in parallel but in separate pipeline. 

cc : @mmitche

Author: I am leaning towards all the validation in the second pipeline.

Reviewer 2: Signing can be run in parallel with all other validation, though it does need to be run after the signing step (.NET 6 does post-build signing).

Reviewer 2: I agree all validation should still be in the second pipeline. There are very few (if any) validation errors where absolutely have to rebuild. They're all cases where we evaluate and make an assessment of the severity of the issue.

The second pipeline Stage-Dotnet-Validate-Publish pipeline downloads the signed build artifacts from the Stage-Dotnet-Sign-Artifacts pipeline. 

One of the main reason for adding the trigger to kick off a the Stage-Dotnet-Validate-Publish from the Stage-Dotnet-Sign-Artifacts pipeline is so that we don't have to manually kick off the Stage-Dotnet-Validate-Publish build after the first pipeline completes. Additionally the Stage-Dotnet-Validate-Publish can be kicked off manually too. It will use BarBuildID / BuildId combination to download the signed assets from the first pipeline. 

## V2:

Rerruning stages in Stage-Dotnet-Validate-Publish pipeline:

(This approach can be re-evaluated later after splitting of the pipeline is completed.)

Say Publishing Signed Assets fails and validation is successful

> 1. If the Signed bits hash has not changed from the previous build, then validation stage is skipped.It directly goes to publishing. 

Reviewer 2: I don't think any of this will need to be automatically determined. The AzDO stage selection feature should allow a CQB to manually select stages in many combinations and get the desired behavior.

Author: Agree, I will change this

Author: Done

  downloadAssets --> post_signing_publishing["Publish Signed \n Assets \n <b>~1h30m</b>"]
  downloadAssets --> vs_insertion["VS Insertion \n Ring \n <b>~50m</b>"]
  vs_insertion --> vs_insertion_override["VS Insertion \n Override \n <b></b>"]
  vs_insertion_override --> test_team_sign_off["Waiting for \n Test team \n to Sign off \n <b></b>"]
  test_team_sign_off --> staging_ring["Staging Ring \n <b></b>"]
  staging_ring --> staging_ring_override["Staging Ring \n Override \n <b></b>"]
  staging_ring_override --> sign_off_for_finalizing_release["Sign off for \nfinalizing \n the release \n <b></b>"]
  sign_off_for_finalizing_release --> finalize_staging_ring["Finalize \n Staging Ring \n <b></b>"]
  sign_off_for_finalizing_release --> publish_cit_validated_assets["Publish CTI \n validated assets \n <b>1h20m</b>"]
  finalize_staging_ring --> approve_publishing_dotnetcs_internal["Approve \n publishing to \n dotnetcsinternal \n <b></b>"]
  approve_publishing_dotnetcs_internal --> handoff_publishing_ring["Handoff \n Publishing Ring \n (dotnetcsinternal) \n <b></b>"]
```

#### Risks: 

> 1. We are disecting the staging pipeline. We need to make sure all the assets are published to the correct storage containers. 

Reviewer 2: ```suggestion
1. We are dissecting the staging pipeline. We need to make sure all the assets are published to the correct storage containers. 
```

> 1. Duplication of asset publishing must be avoided. This will be tracked as a part of this (issue)[https://github.com/dotnet/arcade/issues/13025]

Reviewer 2: A lot of the staging pipeline is already pretty intelligent about this from run to run. In cases where the target location does not need to be stable from run to run (e.g. containers for testing), it will create a unique container name. In cases where the target location needs to be stable run to run, then it verifies that it's not attempting to overwrite artifacts with new ones that have different content.

Author: Agree, we have another effort in the super epic to remove publishing to storage accounts that could be redundant. So I mentioned that instead.

### Splitting Staging Pipeline:

Proposed implementation is splitting the pipeline into two pipelines.
#### 1. First pipeline: Stage-Dotnet-Prepare-Artifacts

> The Stage-Dotnet-Prepare-Artifacts pipeline will contain stages that alters the artifacts. 

Reviewer 2: ```suggestion
The Stage-Dotnet-Prepare-Artifacts pipeline will contain stages that alters or generates new (e.g. SBOM) artifacts. 
```

```

#### Risks: 
1. We are dissecting the staging pipeline, there is a chance we might be messing up publishing to correct storage containers. 

> 1. We are going to retire the old pipeline only after the new pipelines Stage-Dotnet-Prepare-Artifacts and  Stage-Dotnet-Validate-Publish are up and running.

Reviewer 2: ```suggestion
1. Since we will retire the old pipeline only after the new pipelines Stage-Dotnet-Prepare-Artifacts and Stage-Dotnet-Validate-Publish are up and running, there may be an extended period where we need to maintain extra pipelines.
```
